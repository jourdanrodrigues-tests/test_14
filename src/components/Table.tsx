import React from 'react';

type Item = { id: string; [key: string]: unknown };

export type Column<T extends Item> = {
  label: string;
} & (
  | { source: keyof T; render?: never }
  | { source?: never; render: (item: T) => React.ReactNode }
);

type Props<T extends Item> = {
  columns: Column<T>[];
  items: T[];
};

export default function Table<T extends Item>({ columns, items }: Props<T>) {
  return (
    <table className="table-auto relative w-full border">
      <thead>
        <tr className="bg-blue-gray-50 sticky top-0">
          {columns.map(({ label }) => (
            <th key={label} className="px-6 py-3">
              {label}
            </th>
          ))}
        </tr>
      </thead>
      <tbody className="divide-y">
        {items.map((item) => (
          <tr key={item.id} className="even:bg-blue-gray-50/50">
            {columns.map(({ source, render }, index) => (
              <td key={`${item.id}-${index}`} className="py-3 px-4">
                {source && (item[source] as React.ReactNode)}
                {render && render(item)}
              </td>
            ))}
          </tr>
        ))}
      </tbody>
    </table>
  );
}
