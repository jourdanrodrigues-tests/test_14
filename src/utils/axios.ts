import { AxiosError } from 'axios';

const NETWORK_CODES = new Set(['ERR_NETWORK', 'ECONNABORTED']);

export function isOffline(e: unknown) {
  return !!(e instanceof AxiosError && e.code && NETWORK_CODES.has(e.code));
}

export function isServerError(e: unknown): boolean {
  return (
    e instanceof AxiosError && Math.floor((e.response?.status || 0) / 100) === 5
  );
}
