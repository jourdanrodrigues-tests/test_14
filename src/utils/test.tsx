import React from 'react';

import type { UseQueryResult } from '@tanstack/react-query';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render as rtlRender, waitFor } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import { expect } from 'bun:test';
import pretty from 'pretty';

type TQueryHooks<
  TData = unknown,
  TError = unknown,
  QueryResult = UseQueryResult<TData, TError>,
> = Array<{
  useHook(): QueryResult;
  validate(query: QueryResult): boolean;
}>;

type RenderOptions = {
  queryHooks?: TQueryHooks;
};

export const testQueryClient = new QueryClient();
const queryClientWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
);

export async function render(
  node: React.ReactElement,
  options?: RenderOptions
) {
  const hookPromises = (options?.queryHooks || []).map(
    ({ useHook, validate }) => {
      const { result } = renderHook(useHook, { wrapper: queryClientWrapper });
      return waitFor(() => validate(result.current));
    }
  );
  if (hookPromises.length > 0) {
    await Promise.all(hookPromises);
  }

  const result = rtlRender(node, {
    wrapper: ({ children }) => wrapNode(children),
  });
  return result.container;
}

function wrapNode(node: React.ReactNode): React.ReactNode {
  return queryClientWrapper({ children: node });
}

export function expectSnapshot(element: HTMLElement | null) {
  // TODO: Use @testing-library/jest-dom instead of @happy-dom
  //  The JestDOM does some HTML pretty-printing that is better for diffs
  return expect(element && pretty(element.innerHTML));
}
