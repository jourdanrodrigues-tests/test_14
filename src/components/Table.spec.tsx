import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, it } from 'bun:test';

import Table from '@/components/Table.tsx';
import { expectSnapshot, render } from '@/utils/test';

describe('Table Component', () => {
  it('should match the snapshot', async () => {
    const user = userEvent.setup();
    const rows = [
      { id: 'ID_1', name: 'Name 1' },
      { id: 'ID_3', name: 'Name 3' }, // Purposefully out of order
      { id: 'ID_2', name: 'Name 2' },
    ];

    const result = await render(
      <Table
        columns={[
          { label: 'ID', sortBy: 'id', source: 'id' },
          { label: 'Row', source: 'name' },
        ]}
        rowClassName={(_, index) => (index === 1 ? 'custom-row-class' : '')}
        items={rows}
      />
    );

    expectSnapshot(result.container).toMatchSnapshot('unfiltered');
    await user.click(screen.getByText('▲'));
    expectSnapshot(result.container).toMatchSnapshot('ascending');
    await user.click(screen.getByText('▼'));
    expectSnapshot(result.container).toMatchSnapshot('descending');
    await user.click(screen.getByText('▼'));
    expectSnapshot(result.container).toMatchSnapshot('back to unfiltered');
  });
});
