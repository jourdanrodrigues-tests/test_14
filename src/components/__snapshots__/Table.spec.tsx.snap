// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component should match the snapshot: unfiltered 1`] = `
"<table class="table-auto relative w-full border">
  <thead>
    <tr class="bg-blue-gray-50 sticky top-0">
      <th class="px-6 py-3">
        <div class="flex gap-4">
          <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">ID</p>
          <div class="flex flex-col">
            <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▲</p>
            <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▼</p>
          </div>
        </div>
      </th>
      <th class="px-6 py-3">
        <div class="flex gap-4">
          <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">Row</p>
        </div>
      </th>
    </tr>
  </thead>
  <tbody class="divide-y">
    <tr class="even:bg-blue-gray-50/50">
      <td class="py-3 px-4">ID_1</td>
      <td class="py-3 px-4">Name 1</td>
    </tr>
    <tr class="even:bg-blue-gray-50/50">
      <td class="py-3 px-4">ID_3</td>
      <td class="py-3 px-4">Name 3</td>
    </tr>
    <tr class="even:bg-blue-gray-50/50">
      <td class="py-3 px-4">ID_2</td>
      <td class="py-3 px-4">Name 2</td>
    </tr>
  </tbody>
</table>"
`;

exports[`Table Component should match the snapshot: ascending 1`] = `
"<table class="table-auto relative w-full border">
  <thead>
    <tr class="bg-blue-gray-50 sticky top-0">
      <th class="px-6 py-3">
        <div class="flex gap-4">
          <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">ID</p>
          <div class="flex flex-col">
            <p class="block antialiased font-sans font-light text-xs leading-none cursor-pointer select-none text-yellow-700 dark:text-black">▲</p>
            <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▼</p>
          </div>
        </div>
      </th>
      <th class="px-6 py-3">
        <div class="flex gap-4">
          <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">Row</p>
        </div>
      </th>
    </tr>
  </thead>
  <tbody class="divide-y">
    <tr class="even:bg-blue-gray-50/50">
      <td class="py-3 px-4">ID_1</td>
      <td class="py-3 px-4">Name 1</td>
    </tr>
    <tr class="even:bg-blue-gray-50/50">
      <td class="py-3 px-4">ID_2</td>
      <td class="py-3 px-4">Name 2</td>
    </tr>
    <tr class="even:bg-blue-gray-50/50">
      <td class="py-3 px-4">ID_3</td>
      <td class="py-3 px-4">Name 3</td>
    </tr>
  </tbody>
</table>"
`;

exports[`Table Component should match the snapshot: descending 1`] = `
"<table class="table-auto relative w-full border">
  <thead>
    <tr class="bg-blue-gray-50 sticky top-0">
      <th class="px-6 py-3">
        <div class="flex gap-4">
          <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">ID</p>
          <div class="flex flex-col">
            <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▲</p>
            <p class="block antialiased font-sans font-light text-xs leading-none cursor-pointer select-none text-yellow-700 dark:text-black">▼</p>
          </div>
        </div>
      </th>
      <th class="px-6 py-3">
        <div class="flex gap-4">
          <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">Row</p>
        </div>
      </th>
    </tr>
  </thead>
  <tbody class="divide-y">
    <tr class="even:bg-blue-gray-50/50">
      <td class="py-3 px-4">ID_3</td>
      <td class="py-3 px-4">Name 3</td>
    </tr>
    <tr class="even:bg-blue-gray-50/50">
      <td class="py-3 px-4">ID_2</td>
      <td class="py-3 px-4">Name 2</td>
    </tr>
    <tr class="even:bg-blue-gray-50/50">
      <td class="py-3 px-4">ID_1</td>
      <td class="py-3 px-4">Name 1</td>
    </tr>
  </tbody>
</table>"
`;

exports[`Table Component should match the snapshot: back to unfiltered 1`] = `
"<table class="table-auto relative w-full border">
  <thead>
    <tr class="bg-blue-gray-50 sticky top-0">
      <th class="px-6 py-3">
        <div class="flex gap-4">
          <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">ID</p>
          <div class="flex flex-col">
            <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▲</p>
            <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▼</p>
          </div>
        </div>
      </th>
      <th class="px-6 py-3">
        <div class="flex gap-4">
          <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">Row</p>
        </div>
      </th>
    </tr>
  </thead>
  <tbody class="divide-y">
    <tr class="even:bg-blue-gray-50/50">
      <td class="py-3 px-4">ID_1</td>
      <td class="py-3 px-4">Name 1</td>
    </tr>
    <tr class="even:bg-blue-gray-50/50">
      <td class="py-3 px-4">ID_3</td>
      <td class="py-3 px-4">Name 3</td>
    </tr>
    <tr class="even:bg-blue-gray-50/50">
      <td class="py-3 px-4">ID_2</td>
      <td class="py-3 px-4">Name 2</td>
    </tr>
  </tbody>
</table>"
`;
